Mathlib/Data/Nat/Choose/Mchoose.lean

```diff
@@ -0,0 +1,106 @@
+/-
+Copyright (c) 2024 Antoine Chambert-Loir & María-Inés de Frutos—Fernández. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir, María-Inés de Frutos—Fernández
+-/
+
+import Mathlib.Data.Nat.Choose.Vandermonde
+
+/-! # Multiple choose
+
+* `Nat.mchoose m n` is the number of possibilities of dividing a set
+with `m * n` elements into `m` groups of `n`-element subsets
+
+This integer is defined as a finite product.
+
+* `Nat.mchoose_eq` shows that it equals `(m * n).factorial / (m.factorial * n.factorial ^ m)`
+and `Nat.mchoose_mul_factorial_mul_pow_factorial` establishes the equality
+  `mchoose m n * m.factorial * n.factorial ^ m = (m * n).factorial`.
+
+* `Nat.mchoose_mul_add` and `Nat.choose_mul_right` are lemmas for `Nat.choose`
+that couldn't lie in `Mathlib.Data.Nat.Choose.Basic` because their proof uses the `ring` tactic.
+
+-/
+
+namespace Nat
+
+theorem choose_mul_add (m) {n : ℕ} (hn : n ≠ 0) :
+    (m * n + n).choose n = (m + 1) * (m * n + n - 1).choose (n - 1) := by
+  rw [← mul_left_inj' (mul_ne_zero (factorial_ne_zero (m * n)) (factorial_ne_zero n))]
+  set p := n - 1
+  have hp : n = p + 1 := (succ_pred_eq_of_ne_zero hn).symm
+  simp only [hp, add_succ_sub_one]
+  calc
+    (m * (p + 1) + (p + 1)).choose (p + 1) * ((m * (p+1))! * (p+1)!)
+      = (m * (p + 1) + (p + 1)).choose (p + 1) * (m * (p+1))! * (p+1)! := by ring
+    _ = (m * (p+ 1) + (p + 1))! := by rw [add_choose_mul_factorial_mul_factorial]
+    _ = ((m * (p+ 1) + p) + 1)! := by ring_nf
+    _ = ((m * (p + 1) + p) + 1) * (m * (p + 1) + p)! := by rw [factorial_succ]
+    _ = (m * (p + 1) + p)! * ((p + 1) * (m + 1)) := by ring
+    _ = ((m * (p + 1) + p).choose p * (m * (p+1))! * (p)!) * ((p + 1) * (m + 1)) := by
+        rw [add_choose_mul_factorial_mul_factorial]
+    _ = (m * (p + 1) + p).choose p * (m * (p+1))! * (((p + 1) * (p)!) * (m + 1)) := by ring
+    _ = (m * (p + 1) + p).choose p * (m * (p+1))! * ((p + 1)! * (m + 1)) := by rw [factorial_succ]
+    _ = (m + 1) * (m * (p + 1) + p).choose p * ((m * (p + 1))! * (p + 1)!) := by ring
+
+theorem choose_mul_right (m) {n : ℕ} (hn : n ≠ 0) :
+    (m * n).choose n = m * (m * n - 1).choose (n - 1) := by
+  by_cases hm : m = 0
+  · simp only [hm, zero_mul, choose_eq_zero_iff]
+    exact Nat.pos_of_ne_zero hn
+  · set p := m - 1; have hp : m = p + 1 := (succ_pred_eq_of_ne_zero hm).symm
+    simp only [hp]
+    rw [add_mul, one_mul, choose_mul_add _ hn]
+
+/-- Number of possibilities of dividing a set with `m * n` elements into `m` groups
+of `n`-element subsets. -/
+def mchoose (m n : ℕ) : ℕ :=
```

Comment:
Is this different from `Nat.multichoose`?

Reply:
This has nothing to do. Maybe we should find a better name. It is a particular case of the following concept : given a finite set s and a partition p = (p_1,...,p_k) of Card(s), how many (unordered) partitions of the set are there such that its members have sizes p_1,...,p_k. Here we take s of cardinality n * m and p = (n,...,n).

Reply:
(I would likely call them restricted Bell numbers, but that does not seem to be the accepted terminology.)
A definition (using a division) is 
```quote
def rBell (m : Multiset ℕ) :=  (m.sum)! / ((m.map fun n ↦ n !).prod * (∏ n in m.toFinset, (m.count n)!))
```
and then, one has
```
mchoose m n = rBell (Multiset.replicate m n)
```

Reply:
(I have formalized that this works, but this is clearly off topic, and, unfortunately uses the properties of that PR. The difficulty is to prove that the denominator divides the numerator.)

Reply:
To be clear, my comment was primarily "we either need a better name, or documentation to help disambiguate the names, or both". I agree that this is not the same thing!

Reply:
Now we have `Multiset.bell` and `Nat.uniformBell`.